File: src/Binary Tree Level Order Traversal II.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> > levelOrderBottom(TreeNode *root) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
         vector<vector<int> > ret;
        if(!root)
            return ret;
        vector<int> eachlevel;
        int level = 0;
        queue<pair<int, TreeNode*> > q;
        q.push(make_pair(0, root));
        while(!q.empty())
        {
            TreeNode *cur_node = q.front().second;
            int cur_level = q.front().first;
            q.pop();
            if(cur_level == level)
                eachlevel.push_back(cur_node->val);
            else
            {
                level++;
                ret.push_back(eachlevel);
                eachlevel.clear();
                eachlevel.push_back(cur_node->val);
            }
            if(cur_node->left)
                q.push(make_pair(cur_level+1, cur_node->left));
            if(cur_node->right)
                q.push(make_pair(cur_level+1, cur_node->right));
        }
        ret.push_back(eachlevel);
        reverse(ret.begin(), ret.end());
        return ret;
    }
};

File: python/Binary Tree Level Order Traversal II.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of lists of integers
    def levelOrderBottom(self, root):
        ret = []
        if root is None:
            return ret

        level = 0
        curlevel = 0
        nodelist = []
        q = []
        q.append((level, root))
        while len(q) > 0:
            curlevel, node = q.pop(0)
            if curlevel != level:
                level = curlevel
                ret.append(nodelist)
                nodelist = []
            
            nodelist.append(node.val)
            if node.left != None:
                q.append((curlevel+1, node.left))
            if node.right != None:
                q.append((curlevel+1, node.right))

        ret.append(nodelist)
        ret.reverse()
        return ret

